server/src/main/java/org/elasticsearch/common/geo/DimensionalShapeType.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.geo;
+
+import org.apache.lucene.store.ByteArrayDataInput;
+import org.apache.lucene.store.ByteBuffersDataOutput;
+import org.elasticsearch.geometry.Circle;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.GeometryCollection;
+import org.elasticsearch.geometry.GeometryVisitor;
+import org.elasticsearch.geometry.Line;
+import org.elasticsearch.geometry.LinearRing;
+import org.elasticsearch.geometry.MultiLine;
+import org.elasticsearch.geometry.MultiPoint;
+import org.elasticsearch.geometry.MultiPolygon;
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.Polygon;
+import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.geometry.ShapeType;
+
+import java.util.Comparator;
+
+/**
+ * Like {@link ShapeType} but has specific
+ * types for when the geometry is a {@link GeometryCollection} and
+ * more information about what the highest-dimensional sub-shape
+ * is.
+ */
+public enum DimensionalShapeType {
+    POINT,
+    MULTIPOINT,
+    LINESTRING,
+    MULTILINESTRING,
+    POLYGON,
+    MULTIPOLYGON,
+    GEOMETRYCOLLECTION_POINTS,     // highest-dimensional shapes are Points
+    GEOMETRYCOLLECTION_LINES,      // highest-dimensional shapes are Lines
+    GEOMETRYCOLLECTION_POLYGONS;   // highest-dimensional shapes are Polygons
+
+    public static Comparator<DimensionalShapeType> COMPARATOR = Comparator.comparingInt(DimensionalShapeType::centroidDimension);
```

Comment:
I considered it, but thought that type of implicit relationship could be dangerous and unclear to a reader. If you think that concern is overblown, then I am happy to re-use the enum's ordinal here and document the nature of the ordering

Reply:
I think the thing that I found a bit trappy is that this comparator is called just COMPARATOR, which suggests it is the natural order for this enum, yet Java enums also implement Comparable and in that case it would be a different order. I suggested reordering constants, but renaming the comparator would work for me too.

Reply:
makes sense. I agree it isn't done the best way. I'll address this. thanks!

Reply:
out of necessity (found bugs), the whole concept of this COMPARATOR, as well as some of the other values in this enum had to change. I opened #52500 to address the changes that should also resolve this comment